packages/libs/avector/src/lib.rs

```diff
@@ -0,0 +1,551 @@
+//! we mutate and read the array at the same time we can guarantee, that never an uninitialized
+//! **or** write on the same address, entry will take place.
+//! Only once we can guarantee that writing to an entry has been done, and will never take place we
+//! will increment the length.
+//!
+//! Once a bucket has filled up, the vec will allocate a new bucket on the heap and use that,
+//! creating a linked list (into one direction).
+//! This means that every `N`th insertion will be slightly slower, as an allocation needs to be
+//! performed.
+#![no_std]
+#![warn(
+    unreachable_pub,
+    clippy::pedantic,
+    clippy::nursery,
+    clippy::undocumented_unsafe_blocks
+)]
+// TODO: once more stable introduce: warning missing_docs
+#![allow(clippy::module_name_repetitions)]
+#![allow(clippy::redundant_pub_crate)]
+
+extern crate alloc;
+
+use alloc::boxed::Box;
+use core::{cell::UnsafeCell, mem::MaybeUninit, ops::Index, ptr};
+
+use crate::{
+    lock::AtomicLock,
+    sync::{AtomicUsize, Ordering},
+};
+
+mod lock;
+pub(crate) mod sync;
+
+// note(bm): [PERFORMANCE] increase from 16 to 64
+//  push (single/multi): ~10%-40%
+//  iter (singe/multi): ~10%
+pub struct AVec<T, const N: usize = 16> {
+    length: AtomicUsize,
+    lock: AtomicLock,
+
+    // SAFETY: This `UnsafeCell` is not checked by loom, this might seem counterintuitive, but
+    // there's a reason for this.
+    //
+    // The `UnsafeCell` can be accessed mutably and by reference at the same, this is okay, because
+    // the only way to get contents is over `Iter` and `push()`, `push()` ensures through an
+    // `AtomicLock`, that there is only ever a single mutable access at the same time and items
+    // will only be modified at the end. `Iter` meanwhile takes a snapshot of the current length
+    // (stored in an AtomicUsize) and then iterates through all items. `length` will never
+    // decrement, and there is no way that items once pushed can be modified or are ever touched.
+    // The length is only incremented once the push has completed, meaning that the pushed item
+    // will not be accessed until `push()` has finished.
+    //
+    // In theory items could be modified through interior mutability, but that is the case with
+    // "normal" Iterators too.
+    head: UnsafeCell<Bucket<T, N>>,
+}
+
+impl<T> Default for AVec<T, 16> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<T, const N: usize> Drop for AVec<T, N> {
+    fn drop(&mut self) {
+        let length = self.length.load(Ordering::Acquire);
+        let (node, offset) = Self::indices(length);
+
+        let head = self.head.get_mut();
+
+        // SAFETY:
+        // * `drop()` means that no one will access the value anymore
+        // * node and offset are correctly computed from `Self::indices`
+        unsafe { head.uninit(node, offset) };
+    }
+}
+
+// SAFETY: We use `UnsafeCell`, the referred `Bucket` is `Send` if `T` is `Send`
```

Comment:
This is not a safety comment, which argues **why** it is safe to access the struct from two locations at the same time.